### PR TITLE
Simplify Optimizer Parallelism Management

### PR DESCRIPTION
**Pull request by gzm0**

### CODE REVIEW

File: linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/GenIncOptimizer.scala [hunk 1]
```diff
@@ -66,19 +69,19 @@ abstract class GenIncOptimizer private[optimizer] (config: CommonPhaseConfig) {
   private var batchMode: Boolean = false
 
   private var objectClass: Class = _
-  private val classes = CollOps.emptyMap[ClassName, Class]
+  private val classes = new ConcurrentHashMap[ClassName, Class]
```
Comment (sjrd) [inline linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/GenIncOptimizer.scala]: Scala's `mutable.TrieMap` is a fairly specific data structure, which is different from a `ConcurrentHashMap`. Are we sure that this change does not create performance regressions, notably in highly contended cases?

Comment (gzm0) [reply linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/GenIncOptimizer.scala]: I can run some basic benchmarks, but I don't have a beefy machine available (my dev machine only has 4 virtual cores). So it's unlikely we'd discover any serious contention issues.

Comment (gzm0) [reply linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/GenIncOptimizer.scala]: Alternatively, I can just keep `CollOps.Map`. Basically all other cleanups are still doable.

(I am concerned that this sets a precedent of "we are too afraid to change something", which is a dangerous place to be in. But that is a meta-point that allows for longer discussion; my main goal here is to unblock further work on #4315).

Comment (sjrd) [reply linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/GenIncOptimizer.scala]: I'm not opposed to the change. I just noticed that it was a real change and brought it up for consideration. At the end of the day, you are the parallelism expert in that area of the code, so if you think it's no big deal, I'm fine with it.

Comment (gzm0) [reply linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/GenIncOptimizer.scala]: Yeah, but now that you mentioned it, I *do* think we shouldn't just change it :P I'll try to not touch any of the parallelization properties of the code but just clean up a bit, if I'm able to do that, I'll settle for that.

**Comment by gzm0** [comment]
Updated with better history for `GenIncOptimizer` and no need for `ConcurrentHashMap` anymore.

File: linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/GenIncOptimizer.scala [hunk 2]
```diff
@@ -785,13 +829,15 @@ abstract class GenIncOptimizer private[optimizer] (config: CommonPhaseConfig) {
...
-  private[optimizer] abstract class MethodImpl(val owner: MethodContainer,
+  private[optimizer] final class MethodImpl(owner: MethodContainer,
       val methodName: MethodName)
       extends OptimizerCore.MethodImpl with OptimizerCore.AbstractMethodID
       with Unregisterable {
 
     private[this] var _deleted: Boolean = false
 
+    private val bodyAskers = CollOps.emptyMap[MethodImpl, Unit]
+    private val registeredTo = CollOps.emptyMap[Unregisterable, Unit]
```
Comment (sjrd) [inline linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/GenIncOptimizer.scala]: `registeredTo` was previously an `AtomicAcc[Unregisterable], i.e., an `Addable[Unregisterable]`. Shouldn't we keep using the `Addable` abstraction during this refactoring?

**Review by sjrd — CHANGES_REQUESTED** [review]

File: linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/GenIncOptimizer.scala (hunk 2 above)
Comment (gzm0) [reply linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/GenIncOptimizer.scala]: Well, the `Addable` implements parallel traversal which comes at a cost of copying before traversing. So I'm not sure that is better. Also, it seems to me that `registeredTo` really should be a `Set` anyways...

Comment (sjrd) [reply linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/GenIncOptimizer.scala]: Right, that makes sense.

**Comment by gzm0** [comment]
I have addressed all the comments.

**Review by sjrd — APPROVED** [review]

**Pull request merged**